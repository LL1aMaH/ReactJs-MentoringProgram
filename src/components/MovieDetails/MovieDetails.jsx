import React, { memo, useCallback } from 'react';

import { Logo } from '../Logo/Logo';

import poster from '../../../public/Movie_040311_3.jpg'

import styles from '../../../styles/MovieDetails.module.scss'

export const MovieDetails = memo(function Preview({ film }) {
  const onError = useCallback((e) => {
    e.currentTarget.src = '/Movie_040311_3.jpg';
  }, []);
  return (
    <>
      <div
        className={styles.preview}
        style={{
          backgroundImage: 'url(/cover-image.jpg)',
          filter: 'blur(5px)',
        }}
      />
      <div className={styles.previewContent}>
        <div className={styles.buttons}>
          <div className={styles.logo}>
            <Logo />
          </div>
          <div className={styles.info}>
            <img src={film?.poster_path || poster} alt="poster" onError={onError} />
            <div className={styles.infoMovie}>
              <div className={styles.titleWithRating}>
                <h1>{film?.title}</h1>
                {!!film?.vote_average && <div className={styles.circle}>{film.vote_average}</div>}
              </div>
              <p>
                <span>{film?.release_date.slice(0, 4)}</span>
                <span>{film?.runtime && `${film.runtime} min`}</span>
              </p>
              <p className={styles.overview}>{film?.overview}</p>
            </div>
          </div>
        </div>
      </div>
    </>
  );
});
