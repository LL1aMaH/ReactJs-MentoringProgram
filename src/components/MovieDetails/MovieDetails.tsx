import React, { memo, SyntheticEvent, useCallback } from 'react';

import { Logo } from 'Components/Logo';

import { TMovie } from 'src/store/types';
import poster from 'Assets/pictures/Movie_040311_3.jpg';

import styles from './MovieDetails.css';

export type TMovieDetailsProps = {
  film: TMovie | null;
};

export const MovieDetails = memo(function Preview({ film }: TMovieDetailsProps): JSX.Element {
  const onError = useCallback((e: SyntheticEvent<HTMLImageElement, Event>) => {
    e.currentTarget.src = '/src/assets/pictures/Movie_040311_3.jpg';
  }, []);
  return (
    <>
      <div
        className={styles.preview}
        style={{
          backgroundImage: 'url(/src/Assets/pictures/cover-image.jpg)',
          filter: 'blur(5px)',
        }}
      />
      <div className={styles.previewContent}>
        <div className={styles.buttons}>
          <div className={styles.logo}>
            <Logo />
          </div>
          <div className={styles.info}>
            <img src={film?.poster_path || poster} alt="poster" onError={onError} />
            <div className={styles.infoMovie}>
              <div className={styles.titleWithRating}>
                <h1>{film?.title}</h1>
                {!!film?.vote_average && <div className={styles.circle}>{film.vote_average}</div>}
              </div>
              <p>
                <span>{film?.release_date.slice(0, 4)}</span>
                <span>{film?.runtime && `${film.runtime} min`}</span>
              </p>
              <p className={styles.overview}>{film?.overview}</p>
            </div>
          </div>
        </div>
      </div>
    </>
  );
});
