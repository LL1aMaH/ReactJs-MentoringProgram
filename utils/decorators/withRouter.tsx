import React from 'react';

import { MemoryRouter } from 'react-router-dom';

export type TComponentFunc = () => JSX.Element;
export type TDecoratorLocation = {
  pathname?: string;
  search?: string;
};

const withRouter = (
  componentFunc: TComponentFunc,
  location: TDecoratorLocation = { pathname: '/' },
): JSX.Element => {
  return <MemoryRouter initialEntries={[location]}>{componentFunc()}</MemoryRouter>;
};

export default withRouter;
