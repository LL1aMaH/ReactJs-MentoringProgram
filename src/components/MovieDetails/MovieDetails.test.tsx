import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import withRouter from '../../../utils/decorators/withRouter';

import { MovieDetails } from './MovieDetails';

describe('MovieDetails', () => {
  it('should renders correctly', () => {
    const component = render(
      withRouter(() => (
        <MovieDetails
          film={{
            id: 447365,
            title: 'Guardians of the Galaxy Vol. 3',
            release_date: '2020-05-01',
            overview: "The third film based on Marvel's Guardians of the Galaxy.",
            runtime: 100,
            genres: [],
            tagline: '',
            vote_average: 10,
            vote_count: 9,
            poster_path: 'https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg',
            budget: 0,
            revenue: 0,
          }}
        />
      )),
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should renders correctly without film', () => {
    const component = render(withRouter(() => <MovieDetails film={null} />));
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should renders correctly with onError event', () => {
    const { getByAltText } = render(
      withRouter(() => (
        <MovieDetails
          film={{
            id: 447365,
            title: 'Guardians of the Galaxy Vol. 3',
            release_date: '2020-05-01',
            overview: "The third film based on Marvel's Guardians of the Galaxy.",
            runtime: 100,
            genres: [],
            tagline: '',
            vote_average: 10,
            vote_count: 9,
            poster_path: 'https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg',
            budget: 0,
            revenue: 0,
          }}
        />
      )),
    );
    const image: any = getByAltText('poster');
    fireEvent.error(image);
    expect(image.src).toBe('http://localhost/src/assets/pictures/Movie_040311_3.jpg');
  });
});
