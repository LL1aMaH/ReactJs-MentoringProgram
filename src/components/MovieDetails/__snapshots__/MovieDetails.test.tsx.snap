// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="preview"
    style="background-image: url(/src/Assets/pictures/cover-image.jpg); filter: blur(5px);"
  />
  <div
    class="previewContent"
  >
    <div
      class="buttons"
    >
      <div
        class="logo"
      >
        <a
          class="logo"
          href="/"
        >
          <b>
            netflix
          </b>
          roulette
        </a>
      </div>
      <div
        class="info"
      >
        <img
          alt="poster"
          src="https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg"
        />
        <div
          class="infoMovie"
        >
          <div
            class="titleWithRating"
          >
            <h1>
              Guardians of the Galaxy Vol. 3
            </h1>
            <div
              class="circle"
            >
              10
            </div>
          </div>
          <p>
            <span>
              2020
            </span>
            <span>
              100 min
            </span>
          </p>
          <p
            class="overview"
          >
            The third film based on Marvel's Guardians of the Galaxy.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieDetails should renders correctly without film 1`] = `
<DocumentFragment>
  <div
    class="preview"
    style="background-image: url(/src/Assets/pictures/cover-image.jpg); filter: blur(5px);"
  />
  <div
    class="previewContent"
  >
    <div
      class="buttons"
    >
      <div
        class="logo"
      >
        <a
          class="logo"
          href="/"
        >
          <b>
            netflix
          </b>
          roulette
        </a>
      </div>
      <div
        class="info"
      >
        <img
          alt="poster"
          src="[object Object]"
        />
        <div
          class="infoMovie"
        >
          <div
            class="titleWithRating"
          >
            <h1 />
          </div>
          <p>
            <span />
            <span />
          </p>
          <p
            class="overview"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
